### PR TITLE
Add mets-retriever functionality

### PR DESCRIPTION
**Pull request by tw4l**
This PR adds the functionality for `mets-retriever`, a new package and command-line tool for downloading Archivematica METS files from the Archivematica Storage Service. This was developed for a client who intends to post-process their Archivematica METS files regularly and extract technical metadata fields for use in their collection management system.

The `retrieve-mets fetch-all` command is designed to be idempotent by storing the AIP UUIDs of already-retrieved METS files in a local SQLite database, similar to Automation Tools. Only METS files which have not yet been retrieved will be downloaded on subsequent runs by default. A `retrieve-mets fetch-one` command is available to fetch single METS files as desired, without checking the local database first.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  lint:
+    name: "Lint"
+    runs-on: "ubuntu-18.04"
+    steps:
+      - name: "Check out repository"
+        uses: "actions/checkout@v2"
+      - name: "Set up Python"
+        uses: "actions/setup-python@v2"
+        with:
+          python-version: "3.8"
```
Comment (djjuhasz) [inline .github/workflows/test.yml:59]: I'm curious why you only target 3.8 for linting?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  hooks:
+  - id: black
+    args: [--safe, --quiet]
+    language_version: python3
+- repo: https://gitlab.com/pycqa/flake8
+  rev: 3.8.3
+  hooks:
+  - id: flake8
+    language_version: python3
+- repo: https://github.com/pycqa/isort
```
Comment (djjuhasz) [inline .pre-commit-config.yaml:13]: This is awesome - I hate sorting imports 😆 👍

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+repos:
+- repo: https://github.com/ambv/black
+  rev: 18.9b0
+  hooks:
+  - id: black
+    args: [--safe, --quiet]
+    language_version: python3
+- repo: https://gitlab.com/pycqa/flake8
```
Comment (djjuhasz) [inline .pre-commit-config.yaml:8]: Wow, black & flake8 formatting? 😮 Do they not contradict each other sometimes?

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,23 @@
+.PHONY: clean package package-deps package-source package-upload package-wheel
```
Comment (djjuhasz) [inline Makefile]: Okay, I know I'm being totally OCD here, but is there a reason the PHONY list is in a different order than the actual directives? 😖

File: README.md [hunk 5]
```diff
@@ -1,2 +1,151 @@
 # mets-retriever
-CLI tool for bulk downloading Archivematica METS files
+
+## About
+
+CLI tool for bulk downloading [Archivematica][am] METS files.
+
+Unlike Gloria the flat-coated retriever mix, `mets-retriever` is all about fetching.
```
Comment (djjuhasz) [inline README.md:7]: 🤣

File: README.md [hunk 6]
```diff
@@ -1,2 +1,151 @@
...
+
+![Gloria, the flat-coated retriever mix](media/gloria.jpg)
```
Comment (djjuhasz) [inline README.md:9]: ❤️

File: README.md [hunk 7]
```diff
@@ -1,2 +1,151 @@
...
+
+  METS Retriever CLI tool
+
+Options:
+  --version  Show the version and exit.
+  --help     Show this message and exit.
+
+Commands:
+  fetch-all  Fetch all METS files not already retrieved.
+  fetch-one  Fetch single METS file, even if it's already been retrieved.
```
Comment (djjuhasz) [inline README.md:36]: Nicely written, very clear. 👍

File: README.md [hunk 8]
```diff
@@ -1,2 +1,151 @@
...
+```
+
+### `fetch-all`
+
+To fetch all AIP METS files that have not already been retreived, use the
```
Comment (djjuhasz) [inline README.md:41]: Okay one question: *Where* are the METS files fetched too? The current directory?

File: .github/workflows/test.yml [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    strategy:
+      matrix:
+        include:
+          - python-version: "3.6"
+            toxenv: "py36"
+          - python-version: "3.7"
+            toxenv: "py37"
+          - python-version: "3.8"
+            toxenv: "py38"
+          - python-version: "3.9"
```
Comment (djjuhasz) [inline .github/workflows/test.yml:21]: Wow, that's a lot of Python versions to test. 😀

File: mets_retriever/mets_retriever.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+            "sqlite:///{}".format(DATABASE_FILE), echo=False
+        )
+        Session = sqlalchemy.orm.sessionmaker(bind=engine)
+        Base.metadata.create_all(engine)
+        self.session = Session()
+
+    def check_if_aip_in_database(self, uuid):
+        """Check if AIP in database and return boolean."""
+        matching_aip = self.session.query(AIP).filter_by(uuid=uuid).first()
+        if matching_aip:
```
Comment (djjuhasz) [inline mets_retriever/mets_retriever.py]: Can you just do `return bool(matching_aip)` here?

File: .github/workflows/test.yml (hunk 1 above)
Comment (tw4l) [reply .github/workflows/test.yml:59]: It's a bit arbitrary, tbh! The idea is that the linting checks are Python version-independent but it still needs a version of Python to run.

It's also worth noting that this file, the pre-commit config, the Makefile, tox.ini file, etc. were just copy-pasted wholesale from other projects. Hence why I'd like to just make a Cookiecutter template to simplify getting all this stuff into a new repo :)

File: Makefile (hunk 4 above)
Comment (tw4l) [reply Makefile]: Only cause I copy-pasted it! But I'll fix it so it doesn't bother you or the next developer to notice haha (we do tend to be a detail-focused lot!)

File: .pre-commit-config.yaml (hunk 3 above)
Comment (tw4l) [reply .pre-commit-config.yaml:8]: Right? They get along pretty well, Black being really an automatically applied, opinionated superset of rules on top of flake8, but we do set some flake8 rules in the `tox.ini` file's linting config to make sure they play along nicely in those few spots where they do disagree, e.g. with line lengths:

```
[flake8]
exclude = .git, .tox, __pycache__, old, build, dist, txt, .ini
application-import-names = flake8

select = B,C,E,F,W,T4,B9
ignore =
    E501  # Lines are too long.
    W503  # Line break before binary operator.
    E203  # Whitespace before ':'.

import-order-style = pep8
```

File: mets_retriever/mets_retriever.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        am = AMClient(
+            ss_url=self.storage_service_url,
+            ss_user_name=self.storage_service_username,
+            ss_api_key=self.storage_service_api_key,
+        )
+        aips = self._filter_aips(am.aips())
+        for aip in aips:
+            aip_uuid = aip["uuid"]
+            aip_in_database = self.check_if_aip_in_database(aip_uuid)
+            if not aip_in_database:
```
Comment (djjuhasz) [inline mets_retriever/mets_retriever.py]: Could avoid the single use `aip_in_database` var with:

```
aip_uuid = aip["uuid"]
if not self.check_if_aip_in_database(aip_uuid):
...
```

File: setup.py [hunk 12]
```diff
@@ -0,0 +1,25 @@
+from setuptools import find_packages, setup
```
Comment (djjuhasz) [inline setup.py]: Just a note that setuptools discourages the use of setup.py for security reasons:
https://setuptools.pypa.io/en/latest/userguide/quickstart.html#setup-py

**Review by djjuhasz — APPROVED** [review]
Looks good to me @tw4l! I made a few suggestions, but no blockers.

File: README.md (hunk 8 above)
Comment (tw4l) [reply README.md:41]: Explanation added.

File: mets_retriever/mets_retriever.py (hunk 11 above)
Comment (tw4l) [reply mets_retriever/mets_retriever.py]: Good point!

File: mets_retriever/mets_retriever.py (hunk 10 above)
Comment (tw4l) [reply mets_retriever/mets_retriever.py]: `return matching_aip is not None` will do it :)

File: setup.py (hunk 12 above)
Comment (tw4l) [reply setup.py]: Oh, good to know! Things just keep on changing forever, don't they? :laughing:

**Pull request merged**